### ISSUE TITLE
could not find Lazy implicit value of type io.circe.generic.extras.encoding.ConfiguredObjectEncoder

### ISSUE THREAD

**Issue by gavares** [circe/circe#569]
I believe I have a pretty simple and reproducible case that will cause the error above. In `Principal.scala`:
```scala
package test
sealed trait Principal {
  def name: String
}

object Principal {
  case class UserPrincipal(name: String) extends Principal
  case class AgentPrincipal(name: String) extends Principal
}
```

And in `CirceSerializers.scala`
```scala
package test
import cats.syntax.either._
import io.circe._
import io.circe.generic.extras.semiauto._
import io.circe.generic.extras.{Configuration => CirceCfg}
import test.Principal._

object CirceSerializers {
  implicit val circeCfg: CirceCfg = CirceCfg.default.withDiscriminator("kind")

  // --
  // -- Principal Serializers
  // --
  implicit val AgentPrincipalEncoder: ObjectEncoder[AgentPrincipal] = deriveEncoder[AgentPrincipal]
  implicit val AgentPrincipalDecoder: Decoder[AgentPrincipal] = deriveDecoder[AgentPrincipal]

  implicit val UserPrincipalEncoder: ObjectEncoder[UserPrincipal] = deriveEncoder[UserPrincipal]
  implicit val UserPrincipalDecoder: Decoder[UserPrincipal] = deriveDecoder[UserPrincipal]

  implicit val PrincipalEncoder: ObjectEncoder[Principal] = deriveEncoder[Principal]
  implicit val PrincipalDecoder: Decoder[Principal] = {
    Decoder.instance[Principal] { cursor =>
      cursor.downField("kind").as[String].flatMap {
        case "UserPrincipal" => cursor.as[UserPrincipal]
        case "AgentPrincipal" => cursor.as[AgentPrincipal]
        case other => DecodingFailure(s"Could not locate a Principal decoder for json with `kind` value '$other'", Nil).asLeft
      }
    }
  }
}

```

Having these definitions yields the error described in the title of the issue.

If I simply move the definition of Principal and its companion object into `CirceSerializers.scala` before the definition of `object CirceSerializers`, the error goes away.

A repro repository can be checked out here: https://github.com/gavares/circe-error-repro that will generate the error.

**Comment by gavares** [comment]
FWIW, changing from scala 2.11.8 to 2.12.1 also resolves the issue.

**Comment by travisbrown-stripe** [comment]
Hi @gavares, I just tried this on 2.11.12 and it compiles (I confirmed the failure on 2.11.8), which means this is almost certainly a case of SI-7046, which was partially fixed in either 2.11.9 or 2.11.10. Closing now, but please reopen if you can reproduce the error on 2.11.12.
